massa-versioning-worker/src/versioning_factory.rs

```diff
@@ -0,0 +1,405 @@
+        let version = vi_store
+            .0
+            .iter()
+            .rev()
+            .filter(|(vi, vsh)| vi.component == component && vsh.state == state_active)
+            .find_map(|(vi, vsh)| {
+                let res = vsh.state_at(ts, vi.start, vi.timeout);
+                match res {
+                    Ok(MipStateTypeId::Active) => Some(vi.component_version),
+                    _ => None,
+                }
+            })
+            .unwrap_or(0);
+
+        Ok(version)
+    }
+
+    /// Get all versions in 'Active state' for the associated MipComponent
+    fn get_all_best_component_versions(&self) -> Vec<u32> {
+        let component = Self::get_component();
+        let vi_store_ = self.get_versioning_store();
+        let vi_store = vi_store_.0.read();
+
+        let state_active = MipState::active();
+        let versions_iter = vi_store.0.iter().filter_map(|(vi, vsh)| {
+            (vi.component == component && vsh.state == state_active).then_some(vi.component_version)
+        });
+        let versions: Vec<u32> = iter::once(0).chain(versions_iter).collect();
+        versions
+    }
+
+    /// Get all versions (at any state) for the associated MipComponent
+    fn get_all_component_versions(&self) -> BTreeMap<u32, MipStateTypeId> {
+        let component = Self::get_component();
+        let vi_store_ = self.get_versioning_store();
+        let vi_store = vi_store_.0.read();
+
+        let versions_iter = vi_store.0.iter().filter_map(|(vi, vsh)| {
+            (vi.component == component)
+                .then_some((vi.component_version, MipStateTypeId::from(&vsh.state)))
+        });
+        iter::once((0, MipStateTypeId::Active))
+            .chain(versions_iter)
+            .collect()
+    }
+
+    /// Create an object of type Self::Output
+    fn create(
+        &self,
+        args: &Self::Arguments,
+        strategy: Option<FactoryStrategy>,
+    ) -> Result<Self::Output, Self::Error>;
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    use std::collections::BTreeMap;
+
+    use crate::test_helpers::versioning_helpers::advance_state_until;
+    use crate::versioning::{MipInfo, MipStateHistory};
+
+    use massa_time::MassaTime;
+
+    // Define a struct Address with 2 versions AddressV0 & AddressV1
+    #[derive(Debug)]
+    struct AddressV0 {
+        hash: String,
+    }
+
+    impl AddressV0 {
+        fn new(hash: String) -> Self {
+            Self { hash }
+        }
+    }
+
+    #[derive(Debug)]
+    struct AddressV1 {
+        slot: String,
+        creator: String,
+        index: u32,
+    }
+
+    impl AddressV1 {
+        fn new(slot: String, creator: String, index: u32) -> Self {
+            Self {
+                slot,
+                creator,
+                index,
+            }
+        }
+    }
+
+    #[derive(Debug)]
+    enum Address {
+        V0(AddressV0),
+        V1(AddressV1),
+    }
+
+    //
+    struct AddressArgs {
```

Comment:
Why this is not an enum with `V0Args()` and `V1Args()` variant to avoid optional and regroup everything like that ?

Reply:
The goal here was to keep the unit test code easy to understand and clear but at the end the dev that will implement AddressArgs is free to do whatever he wants to :)

Reply:
Okok